app/include/zmk/event_manager.h

```diff
@@ -68,17 +71,22 @@ struct zmk_event_subscription {
             .listener = &zmk_listener_##mod,                                                       \
     };
 
-#define ZMK_EVENT_RAISE(ev) zmk_event_manager_raise((zmk_event_t *)ev);
+#define ZMK_EVENT_RAISE(ev)                                                                        \
+    ((__ASSERT((uint8_t *)&(ev).header - (uint8_t *)&ev == 0,                                      \
+               "header must be first element of event")),                                          \
+     zmk_event_manager_raise(&(ev).header));
 
 #define ZMK_EVENT_RAISE_AFTER(ev, mod)                                                             \
-    zmk_event_manager_raise_after((zmk_event_t *)ev, &zmk_listener_##mod);
+    ((__ASSERT((uint8_t *)&(ev).header - (uint8_t *)&ev == 0,                                      \
+               "header must be first element of event")),                                          \
```

Comment:
I recommend creating a macro like
```c
#define ZMK_ASSERT_IS_EVENT(ev) \
    (__ASSERT((uint8_t *)&(ev).header - (uint8_t *)&ev == 0,                                      \
               "header must be first element of event"))
```
to simplify this. (Feel free to come up with a better name for this, as "is event" isn't strictly accurate since it only checks the header member, not the type.)